### ISSUE TITLE
pytest: do not store .pytest_cache in $out

### ISSUE THREAD

**Issue by FRidh** [NixOS/nixpkgs#40273]
## Issue description

In #40259 a collision of `.pytest_cache` files was reported. These files should not exist in the first place. 

They only seem to be introduced when running `pytest` with `$out` as target, so

```
pytest $out
```

**Comment by FRidh** [comment]
Occurrences:
```sh
$ git grep 'py.test $out'
pkgs/development/python-modules/autobahn/default.nix:    py.test $out
pkgs/development/python-modules/cairocffi/default.nix:    py.test $out/${python.sitePackages}
pkgs/development/python-modules/hyperlink/default.nix:    py.test $out
pkgs/development/python-modules/pandas/default.nix:    py.test $out/${python.sitePackages}/pandas --skip-slow --skip-network \
pkgs/development/python-modules/pyzmq/default.nix:    py.test $out/${python.sitePackages}/zmq/ -k "not test_socket \
pkgs/development/python-modules/spake2/default.nix:    py.test $out
pkgs/development/python-modules/xarray/default.nix:    py.test $out/${python.sitePackages}
pkgs/top-level/python-packages.nix:      py.test $out/${python.sitePackages}
pkgs/top-level/python-packages.nix:      py.test $out
pkgs/top-level/python-packages.nix:      py.test $out
pkgs/top-level/python-packages.nix:      py.test $out/${python.sitePackages}
```

**Comment by FRidh** [comment]
Quoting from https://github.com/NixOS/nixpkgs/issues/40259#issuecomment-388009783:

> Just adding for searchability/reference: et_xmlfile still includes this, and pytest could alternatively be called like py.test -p no:cacheprovider $out if we don't want this cache dir created at all.

**Comment by FRidh** [comment]
Fix in https://github.com/NixOS/nixpkgs/commit/610485faa7f8c043482c4a9c030938ed315557e5

**Comment by FRidh** [comment]
Seems like the issue is still present (https://github.com/NixOS/nixpkgs/issues/44479):
```
nix-shell -p 'python3.withPackages(ps: with ps; [ cairocffi tinycss ])' -I nixpkgs=.
```

**Comment by FRidh** [comment]
Looks like a typo; it should be `.pytest_cache` and not `.pytest-cache`.

**Comment by lopsided98** [comment]
This doesn't seem to fix the problem, because `fixupPhase` is run before `installCheckPhase`, which is where python packages run their tests.

**Comment by FRidh** [comment]
It is indeed the incorrect phase, I forgot about the order. The solution is to use `postCheckHooks` instead. However, it is important that when `checkPhase` is overridden, the `postHook` is still ran. Here we could use https://github.com/NixOS/rfcs/pull/32.

**Comment by FRidh** [comment]
I've pushed the fixes to `staging-18.09`. Now we need to wait for that to land in `release-18.09`.

**Comment by matthewbauer** [comment]
Looks to now have landed.

**Comment by alexeymuranov** [comment]
@matthewbauer , are you sure?  I do not see `staging-18.09` merged into `release-18.09` since a while...

**Comment by alexeymuranov** [comment]
The related #40273 is still an issue on 18.09.
